### PR TITLE
Fallback to "en" locale if no file exists in Parsley i18n folder

### PR DESCRIPTION
**Pull request by New0**
For #3295 

Fallback to "en" if no file in assets/js/i18n/ folder matches the locale ( we need to enqueue a file to be sure setLocale() is used for Parsley and it doesn't eturn an error.

@Shelob9 do you have an idea how we could test this ? Consequences are a logged error if no locale was set for Parsley or a 404 response for a missing enqueued asset.

### CODE REVIEW

**Comment by Shelob9** [comment]
Isn't there a filter in WordPress for default language?

Also, we really should have a French test site and a test site in a language we don't have translation files running the acceptance tests.

File: classes/render/assets.php [hunk 1]
```diff
@@ -1081,6 +1080,10 @@ protected static function set_locale_code($locale)
 			$locale = strtolower(substr($locale, 0, 2));
 		} elseif (file_exists(CFCORE_PATH . 'assets/js/i18n/' . strtolower(substr($locale, 3)) . '.js')) {
 			$locale = strtolower(substr($locale, 3));
+		} else {
+			//No file is matching the locale in validation folder, 
+			//Fallback to english instead of not enqueueing for Parsley to be set a locale and not return an error
```
Comment (Shelob9) [inline classes/render/assets.php]: Me: It smells bad that there is nothing in the else, but nothing is wrong.

Also me: It smelled bad for a reason.

File: tests/test-locale-assets-enqueued.php [hunk 2]
```diff
@@ -0,0 +1,50 @@
...
+     */
+    public function testKnownLocale()
+    {
+        
+        $locale = "fr";
+        $code =   Caldera_Forms_Render_Assets::set_locale_code($locale);
+        $validator_url = Caldera_Forms_Render_Assets::get_validator_locale_url($locale);
+
+        $this->assertSame( $code, "fr" );
+        $this->assertSame( $validator_url, "http://example.org/wp-content/plugins/app/assets/js/i18n/fr.js" );
```
Comment (Shelob9) [inline tests/test-locale-assets-enqueued.php]: @New0 I would be careful about assuming site_url. Maybe better to check if the string ends with `i18n/fr.js`.

File: tests/test-locale-assets-enqueued.php [hunk 3]
```diff
@@ -0,0 +1,50 @@
...
+
+     /**
+      * Test a locale which corresponding file doesn't exist within assets/js/i18n/
+      * 
+      * @since 1.8.7
+      *
+      * @covers  \Caldera_Forms_Render_Assets::set_locale_code($locale)
+      * @covers  \Caldera_Forms_Render_Assets::get_validator_locale_url($locale)
+     */
+    public function testUnkKnownLocale()
```
Comment (Shelob9) [inline tests/test-locale-assets-enqueued.php]: @New0 Type in method name

File: tests/test-locale-assets-enqueued.php [hunk 4]
```diff
@@ -0,0 +1,50 @@
...
+    {
+        
+        $locale = "bo";
+        $code =   Caldera_Forms_Render_Assets::set_locale_code($locale);
+        $validator_url = Caldera_Forms_Render_Assets::get_validator_locale_url($locale);
+
+        $this->assertSame( $code, "en" );
+        $this->assertSame( $validator_url, "http://example.org/wp-content/plugins/app/assets/js/i18n/en.js" );
```
Comment (Shelob9) [inline tests/test-locale-assets-enqueued.php]: @New0 Same point as in other test about only testing the end.

File: classes/render/assets.php [hunk 5]
```diff
@@ -1049,14 +1049,14 @@ protected static function field_script_dependencies()
...
 	 * @return string
 	 */
-	protected static function get_validator_locale_url($locale)
+	public static function get_validator_locale_url($locale)
 	{
-		if(!empty($locale)){
-			$locale = self::set_locale_code($locale);
-		}
+		//Use correct formatted locale for file to be matched
+		$locale = self::set_locale_code($locale);
```
Comment (mrpritchett) [inline classes/render/assets.php]: Why are we removing a check here?

File: classes/render/assets.php (hunk 1 above)
Comment (mrpritchett) [reply classes/render/assets.php]: Agree

File: classes/render/assets.php (hunk 5 above)
Comment (New0) [reply classes/render/assets.php]: @mrpritchett The check was removed to always run set_locale_url() which sets a default $locale to "en" if no locale was set or no file matching local found

File: tests/test-locale-assets-enqueued.php [hunk 6]
```diff
@@ -0,0 +1,50 @@
...
+        
+        $locale = "bo";
+        $code =   Caldera_Forms_Render_Assets::set_locale_code($locale);
+        $validator_url = Caldera_Forms_Render_Assets::get_validator_locale_url($locale);
+
+        $this->assertSame( $code, "en" );
+        $this->assertSame( substr($validator_url, -10), "i18n/en.js" );
+    }
+
+}
```
Comment (mrpritchett) [inline tests/test-locale-assets-enqueued.php]: New line.

**Comment by mrpritchett** [comment]
Approved pending new line.

**Pull request merged**